### PR TITLE
DP-188: ruby 2.7

### PR DESCRIPTION
**Pull request by zhelyan**

### CODE REVIEW

**Comment by simonwgill** [comment]
If you add the TargetRubyVersion to the .rubocop.yml, that ought to catch the deprecation warnings, right?

**Comment by zhelyan** [comment]
@simonwgill I tried that but CASG targets 2.6. Rubocop complains that we are using different target versions and doesn't seem to do the extra 2.7 checks

**Comment by simonwgill** [comment]
> @simonwgill I tried that but CASG targets 2.6. Rubocop complains that we are using different target versions and doesn't seem to do the extra 2.7 checks

Did you try it after merging develop in? It seems to be working ok for me, although it isn't catching any new issues

**Comment by zhelyan** [comment]
it does not fail but it prints `required_ruby_version (2.7, declared in cfn.gemspec) and TargetRubyVersion (2.6, which may be specified in .rubocop.yml) should be equal.`

**Comment by simonwgill** [comment]
> it does not fail but it prints `required_ruby_version (2.7, declared in cfn.gemspec) and TargetRubyVersion (2.6, which may be specified in .rubocop.yml) should be equal.`

Yeah, so setting TargetRubyVersion to 2.7 won't be a problem and should enable the right cops. If it's picking up the deprecation warnings without adding the Target, then that's fine too.

**Comment by simonwgill** [comment]
Doesn't seem to matter anyway. This code doesn't have any rubocop errors whether or not the target ruby version is set to 2.7

```ruby
# frozen-string-literal: true

# This method accepts only a keyword argument
def foo(kop: 1)
  p kop
end

h = { kop: 42 }

# This method call passes a positional Hash argument
# In Ruby 2.7: The Hash is automatically converted to a keyword argument
# In Ruby 3.0: This call raises an ArgumentError
foo(h)
# => demo.rb:11: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
#    demo.rb:2: warning: The called method `foo' is defined here
#    42

# If you want to keep the behavior in Ruby 3.0, use double splat
foo(**h) #=> 42
```

File: Dockerfile.dev [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM ruby:2.6.5-alpine
+FROM ruby:2.7-alpine
```
Comment (simonwgill) [inline Dockerfile.dev:1]: This should be 2.7.1 to match the tool version.

**Review by simonwgill — DISMISSED** [review]
Just the Dockerfile change needed

**Comment by zhelyan** [comment]
fair enough

File: Dockerfile.dev (hunk 1 above)
Comment (zhelyan) [reply Dockerfile.dev:1]: Do we plan to update to 2.7.* or we'll jump straight to 3? I assumed the first. Leaving the version in docker at `2.7` could help identify problems with future ruby patch releases early.

Comment (simonwgill) [reply Dockerfile.dev:1]: We probably will be following patch releases, but shouldn't we be matching those to the tool-version in use so we don't get any weird differences between our local machines and the container?

Comment (zhelyan) [reply Dockerfile.dev:1]: yes, we should but that's a separate thing. The pin to 2.7 in the jenkinsfile means the build will use the latest available patch version and builds will inform us about any issues with it: failing build ( unlikely), failing tests ( unlikely ) or new deprecation warnings ( likely ). 
We can then decide whether it's worth updating to the new patch release which will require new pins in .tool-versions and couple of other places.

**Review by ruthwells — DISMISSED** [review]

**Review by michelesr — APPROVED** [review]

**Review by ruthwells — APPROVED** [review]

**Review by simonwgill — APPROVED** [review]

**Pull request merged**